src/NumericalAlgorithms/LinearSolver/LinearSolver.hpp

```diff
@@ -0,0 +1,242 @@
+    if constexpr (not std::is_same_v<Preconditioner, NoPreconditioner>) {
+      preconditioner_ = rhs.clone_preconditioner();
+    }
+  }
+  PreconditionedLinearSolver& operator=(
+      const PreconditionedLinearSolver& rhs) noexcept {
+    Base::operator=(rhs);
+    if constexpr (not std::is_same_v<Preconditioner, NoPreconditioner>) {
+      preconditioner_ = rhs.clone_preconditioner();
+    }
+    return *this;
+  }
+
+ public:
+  ~PreconditionedLinearSolver() override = default;
+
+  /// \cond
+  explicit PreconditionedLinearSolver(CkMigrateMessage* m) noexcept : Base(m) {}
+  /// \endcond
+
+  void pup(PUP::er& p) noexcept override {  // NOLINT
+    PUP::able::pup(p);
+    if constexpr (not std::is_same_v<Preconditioner, NoPreconditioner>) {
+      p | preconditioner_;
+    }
+  }
+
+  /// Whether or not a preconditioner is set
+  bool has_preconditioner() const noexcept {
+    if constexpr (not std::is_same_v<Preconditioner, NoPreconditioner>) {
+      return preconditioner_.has_value();
+    } else {
+      return false;
+    }
+  }
+
+  // @{
+  /// Access to the preconditioner. Check `has_preconditioner()` before calling
+  /// this function. Calling this function when `has_preconditioner()` returns
+  /// `false` is an error.
+  template <bool Enabled = not std::is_same_v<Preconditioner, NoPreconditioner>,
+            Requires<Enabled> = nullptr>
+  const Preconditioner& preconditioner() const noexcept {
+    ASSERT(has_preconditioner(),
+           "No preconditioner is set. Please use `has_preconditioner()` to "
+           "check before trying to retrieve it.");
+    if constexpr (std::is_abstract_v<Preconditioner>) {
+      return **preconditioner_;
+    } else {
+      return *preconditioner_;
+    }
+  }
+
+  template <bool Enabled = not std::is_same_v<Preconditioner, NoPreconditioner>,
+            Requires<Enabled> = nullptr>
+  Preconditioner& preconditioner() noexcept {
+    ASSERT(has_preconditioner(),
+           "No preconditioner is set. Please use `has_preconditioner()` to "
+           "check before trying to retrieve it.");
+    if constexpr (std::is_abstract_v<Preconditioner>) {
+      return **preconditioner_;
+    } else {
+      return *preconditioner_;
+    }
+  }
+
+  /// Copy the preconditioner. Useful when it has an abstract type.
+  template <bool Enabled = not std::is_same_v<Preconditioner, NoPreconditioner>,
+            Requires<Enabled> = nullptr>
+  std::optional<PreconditionerType> clone_preconditioner() const noexcept {
+    if constexpr (std::is_abstract_v<Preconditioner>) {
+      return has_preconditioner()
+                 ? std::optional((*preconditioner_)->get_clone())
+                 : std::nullopt;
+    } else {
+      return preconditioner_;
+    }
+  }
+
+  // Keep the function virtual so derived classes must provide an
+  // implementation, but also provide an implementation below that derived
+  // classes can use to reset the preconditioner
+  void reset() noexcept override = 0;
```

Comment:
I don't remember the rules, but are you sure you can't just not define this and then this class immediately remains an abstract base class?

Reply:
I have to put this line here so I can provide an out-of-line implementation that derived classes can use

Reply:
Sorry, I don't understand what you mean with derived classes implement get_clone. Who and where will the definition be? I'm worried that this will be an ODR violation: you have multiple different definitions for the same function.

Reply:
The `Gmres` class implements `reset` and calls `Base::reset()`, which refers to the implementation in `PreconditionedLinearSolver`. That should be fine right?

`get_clone` is another story. `LinearSolver` declares it virtual and `Gmres` implements it by calling the constructor. The copy constructor calls its base, which is `PreconditionedLinearSolver`. And `PreconditionedLinearSolver` uses this helper-function `clone_preconditioner()` to clone the object it holds in its member variable. Sounds ok?